### DISCUSSION TITLE
File Based Config

### DISCUSSION THREAD

**Discussion by coffee-cup** [railwayapp/nixpacks#120] [Ideas]
Original issue: https://github.com/railwayapp/nixpacks/issues/27

Nixpacks configuration should be able to be provided through a file on disk in the app source. The main benefit is that build configuration can be tracked through source control.

At the moment, the configurable options are

- Nixpkgs to install
- Build command
- Start command

A very basic format for these options is a JSON file

```json
{
  "packages": ["pkgs.cowsay", "pkgs.lolcat"],
  "startCommand": "...",
  "buildCommand": "..."
}
```

Screenshot of comments on original issue 👇

![image](https://user-images.githubusercontent.com/3044853/168215383-f7a32b9b-aea4-45e1-9d62-151ab188979f.png)

**Comment by josegonzalez** [comment]
This would be cool because then you could say "Any app that has a `nixpacks.json` file is a nixpacks app".

**Comment by datner** [comment]
This is an amazing idea!

**Comment by aleksrutins** [comment]
This has theoretically been [implemented as of a while ago](https://nixpacks.com/docs/configuration/file); however, if there's something else folks would like to see, we're open to feedback.
